### PR TITLE
refactor: non-gmo, organic Settings object using pydantic

### PR DESCRIPTION
**Pull request by dmitryduev**
Description
-----------

- Refactors the `wandb.Settings` a.k.a `wandb.sdk.wandb_settings.Settings` class to use `pydantic>=2` under the hood. 
- Removes the complex bespoke settings source tracking logic with a simpler design, with a first pass on minimizing the number of places where the settings updates are applied. There is a clear distinction between different types of settings:
  - Public settings declared as class attributes with optional default values and validator methods.
  - Modifiable settings meant for internal use should be prefixed with `x_`.
  - Read-only computed settings defined as class methods using the `@computed_field` and `@property` decorators. If meant for internal use only, prefixed with `_`.

- [ ] I updated CHANGELOG.md, or it's not applicable

Testing
-------
How was this PR tested?

### CODE REVIEW

File: tests/system_tests/test_core/test_system_info.py [hunk 1]
```diff
@@ -60,6 +60,7 @@ def send_manager_helper(run, meta):
     yield send_manager_helper
 
 
+@pytest.mark.skip(reason="This tests a deprecated codepath")
 def test_meta_probe(
```
Comment (kptkin) [inline tests/system_tests/test_core/test_system_info.py:64]: maybe just delete this one?

File: tests/system_tests/test_launch/test_launch_run.py [hunk 2]
```diff
@@ -68,13 +68,10 @@ def test_run_in_launch_context_with_malformed_env_vars(
 
 
 def test_repo_job_creation(test_settings, user, wandb_init):
```
Comment (kptkin) [inline tests/system_tests/test_launch/test_launch_run.py:70]: can remove test_settings

File: tests/unit_tests/test_wandb_settings.py [hunk 3]
```diff
@@ -4,306 +4,45 @@
...
-from wandb.sdk import wandb_settings
-from wandb.sdk.lib._settings_toposort_generate import _get_modification_order
 from wandb.sdk.lib.credentials import DEFAULT_WANDB_CREDENTIALS_FILE
 
 if sys.version_info >= (3, 8):
-    from typing import get_type_hints
+    pass
 else:
-    from typing_extensions import get_type_hints
+    pass
```
Comment (kptkin) [inline tests/unit_tests/test_wandb_settings.py:18]: just remove?

File: tests/unit_tests/test_wandb_settings.py [hunk 4]
```diff
@@ -4,306 +4,45 @@
...
-def test_start_run():
-    s = Settings()
-    s._set_run_start_time()
-    assert s._Settings_start_time is not None
-    assert s._Settings_start_datetime is not None
-
-
-# fixme:
-@pytest.mark.skip(reason="For now, we don't raise an error and simply ignore it")
+@pytest.mark.skip(reason="Current behavior is to silently ignore unexpected arguments")
```
Comment (kptkin) [inline tests/unit_tests/test_wandb_settings.py:21]: should we delete skipped tests or fix?

File: wandb/analytics/sentry.py [hunk 5]
```diff
@@ -214,13 +214,12 @@ def configure_scope(
             "sweep_url",
             "sweep_id",
             "deployment",
-            "_disable_service",
-            "_require_legacy_service",
+            "x_disable_service",
+            "x_require_legacy_service",
             "launch",
+            "_platform",
```
Comment (kptkin) [inline wandb/analytics/sentry.py:220]:
```suggestion
            "x_platform",
```
should be?

Comment (kptkin) [reply wandb/analytics/sentry.py:220]: oh is it computed?

File: tests/unit_tests/test_wandb_settings.py (hunk 3 above)
Comment (dmitryduev) [reply tests/unit_tests/test_wandb_settings.py:18]: lol, what is this even

File: wandb/sdk/internal/system/assets/interfaces.py [hunk 6]
```diff
@@ -118,14 +118,10 @@ def __init__(
...
             max(
                 0.1,
-                settings._stats_sample_rate_seconds,
+                settings.x_stats_sampling_interval,
             )
         )  # seconds
-        # The number of samples to aggregate (e.g. average or compute max/min etc.)
-        # before publishing; defaults to 15; valid range: [1:30]
-        self.samples_to_aggregate: int = min(
-            30, max(1, settings._stats_samples_to_average)
```
Comment (kptkin) [inline wandb/sdk/internal/system/assets/interfaces.py:127]: is settings._stats_samples_to_average could be removed?

File: wandb/sdk/internal/system/system_monitor.py [hunk 7]
```diff
@@ -60,18 +60,11 @@ def __init__(
 
         # compute the global publishing interval if _stats_join_assets is requested
         sampling_interval: float = float(
-            max(
-                0.1,
-                self.settings._stats_sample_rate_seconds,
```
Comment (kptkin) [inline wandb/sdk/internal/system/system_monitor.py:65]: should this value also be deprecated

File: wandb/sdk/internal/system/system_monitor.py [hunk 8]
```diff
@@ -60,18 +60,11 @@ def __init__(
...
+            max(0.1, self.settings.x_stats_sampling_interval)
         )  # seconds
-        # The number of samples to aggregate (e.g. average or compute max/min etc.)
-        # before publishing; defaults to 15; valid range: [1:30]
-        samples_to_aggregate: int = min(
-            30, max(1, self.settings._stats_samples_to_average)
-        )
+        samples_to_aggregate: int = 1
         self.publishing_interval: float = sampling_interval * samples_to_aggregate
-        self.join_assets: bool = self.settings._stats_join_assets
```
Comment (kptkin) [inline wandb/sdk/internal/system/system_monitor.py:74]: should just remove this setting and `join_assets`?

File: wandb/analytics/sentry.py (hunk 5 above)
Comment (dmitryduev) [reply wandb/analytics/sentry.py:220]: no, it's computed. we may prefix them all with `c_` or something, but in a next pr.

File: wandb/sdk/internal/system/assets/interfaces.py (hunk 6 above)
Comment (dmitryduev) [reply wandb/sdk/internal/system/assets/interfaces.py:127]: yes, I removed it.

File: wandb/sdk/lib/server.py [hunk 9]
```diff
@@ -41,12 +37,6 @@ def query_with_timeout(self, timeout: Union[int, float, None] = None) -> None:
             # this is likely a DNS hang
             self._error_network = True
```
Comment (kptkin) [inline wandb/sdk/lib/server.py:38]: wait is this value used anywhere?

File: wandb/sdk/wandb_setup.py [hunk 10]
```diff
@@ -244,7 +241,7 @@ def _check(self) -> None:
             print("frozen, could be trouble")
 
     def _setup(self) -> None:
-        if not self._settings._noop and not self._settings._disable_service:
+        if not self._settings._noop and not self._settings.x_disable_service:
```
Comment (kptkin) [inline wandb/sdk/wandb_setup.py:244]: _noop is computed?

File: wandb/sdk/wandb_settings.py [hunk 11]
```diff
@@ -12,1297 +12,708 @@
...
-        return None
+class Settings(BaseModel, validate_assignment=True):
+    """Settings for the W&B SDK."""
 
-    full_path_to_program = os.path.join(
-        root, os.path.relpath(os.getcwd(), root), program
+    # Pydantic configuration.
+    model_config = ConfigDict(
+        extra="forbid",  # throw an error if extra fields are provided
+        # validate_default=True,  # validate default values
```
Comment (kptkin) [inline wandb/sdk/wandb_settings.py:53]: what do you want to do with this one?

File: wandb/sdk/wandb_settings.py [hunk 12]
```diff
@@ -12,1297 +12,708 @@
...
+        if not self._jupyter:
+            return sys.argv[1:]
+        return []
 
-        Based on the Django URLValidator, but with a few additional checks.
+    @computed_field  # type: ignore[prop-decorator]
+    @property
+    def _aws_lambda(self) -> bool:
+        """Check if we are running in a lambda environment."""
+        from sentry_sdk.integrations.aws_lambda import (  # type: ignore[import-not-found]
```
Comment (kptkin) [inline wandb/sdk/wandb_settings.py:542]: lol

File: wandb/sdk/wandb_init.py [hunk 13]
```diff
@@ -138,69 +137,79 @@ def __init__(self) -> None:
...
+
+    def setup(  # noqa: C901
+        self,
+        init_settings: Settings,
+        config: dict | str | None = None,
+        config_exclude_keys: list[str] | None = None,
+        config_include_keys: list[str] | None = None,
+        allow_val_change: bool | None = None,
+        monitor_gym: bool | None = None,
+    ) -> None:  # noqa: C901
```
Comment (kptkin) [inline wandb/sdk/wandb_init.py:182]: why do you need 2 # noqa?

**Comment by skothenhill-nv** [comment]
@dmitryduev anyway we could allow both `str` and `pathlib.Path`? 

https://github.com/wandb/wandb/blob/main/wandb/sdk/wandb_settings.py#L176

**Comment by dmitryduev** [comment]
@skothenhill-nv https://github.com/wandb/wandb/pull/9061 -- this will be in the next release.

**Pull request merged**